src/nlp/arrow_writer.py

```diff
@@ -209,49 +221,74 @@ def __init__(
 
         self._path = path
+        self._parquet_path = os.path.splitext(path)[0] + ".parquet"
+        self._namespace = namespace or "default"
         self._num_examples = None
-        self._pcoll_outputs_metadata = []
+        self._cache_dir = cache_dir or HF_DATASETS_CACHE
 
     def write_from_pcollection(self, pcoll_examples):
+        """Add the final steps of the beam pipeline: write to parquet files."""
         import apache_beam as beam
+        from .utils.beam_utils import WriteToParquet
 
-        # create some metadata that will be used in .finalize()
-        num_examples = (
-            pcoll_examples
-            | "Add metadata key" >> beam.Map(lambda v: ("num_examples", v))
-            | "Count" >> beam.CombinePerKey(beam.transforms.combiners.CountCombineFn())
-        )
-
-        def save_metatada(metadata_items):
-            with open(self._path + ".json", "w") as metadata_file:
-                json.dump(metadata_items, metadata_file)
+        def inc_num_examples(example):
+            beam.metrics.Metrics.counter(self._namespace, "num_examples").inc()
 
-        # save metadata
-        _ = (
-            (num_examples,)
-            | "Merge pcollections" >> beam.Flatten()
-            | "Create Dict" >> beam.transforms.combiners.ToDict()
-            | "Save metadata" >> beam.ParDo(save_metatada)
-        )
+        # count examples
+        _ = pcoll_examples | "Count N. Examples" >> beam.Map(inc_num_examples)
 
         # save dataset
         return (
             pcoll_examples
             | "Get values" >> beam.Values()
             | "Save to parquet"
-            >> beam.io.parquetio.WriteToParquet(self._path, self._schema, num_shards=1, shard_name_template="")
+            >> WriteToParquet(self._parquet_path, self._schema, num_shards=1, shard_name_template="")
         )
 
-    def finalize(self):
-        self._num_bytes = os.path.getsize(self._path)
-        with open(self._path + ".json", "r") as metadata_file:
-            metadata = json.load(metadata_file)
-        self._num_examples = metadata["num_examples"]
-        os.remove(self._path + ".json")
-        logger.info(
-            "Done writing %s examples in %s bytes %s.",
-            self._num_examples,
-            self._num_bytes,
-            self._path if self._path else "",
-        )
+    def finalize(self, metrics_query_result: dict):
+        """
+        Run after the pipeline has finished.
+        It converts the resulting parquet files to arrow and it completes the info from the pipeline metrics.
+
+        Args:
+            metrics_query_result: `dict` obtained from pipeline_results.metrics().query(m_filter). Make sure
+                that the filter keeps only the metrics for the considered split, under the namespace `split_name`.
+        """
+        import apache_beam as beam
+        from .utils import beam_utils
+
+        # Convert to arrow
```

Comment:
Here os the conversion from parquet to arrow after each pipeline run

Reply:
By default the conversion is done by streaming data like this:
parquet file on gsc -> local machine -> arrow file one gcs

However if you have a bad connection (as I do), it might happen that the socket connection fails (if you have a stabke connection it should happen). In this case I do the basic download locally + convert + upload...

Reply:
I couldn't do it directly inside the Beam pipeline because writing to files is done asynchronously and there is no way in python to know when the parquet file writing has finished (apparently the Java sdk does but not python...)

Reply:
I am thinking of using something else that the local machine to do it. We could have a separate beam pipeline just for conversion, and this pipeline could simply be run after the previous one have finished. Or we could use any other tool...